File: nodejs/packages/periph/src/chips/power/ade7953.js
'use strict';

// 8-bit registers
const REG_DISNOLOAD       = 0x001;
const REG_PGA_V           = 0x007;
const REG_PGA_IA          = 0x008;
const REG_PGA_IB          = 0x009;
const REG_WRITE_PROTECT   = 0x040;
const REG_VERSION         = 0x702;
const REG_EX_REF          = 0x800;

// 16-bit registers
const REG_CONFIG          = 0x102;
const REG_CF1DEN          = 0x103;
const REG_CF2DEN          = 0x104;
const REG_CFMODE          = 0x107;
const REG_PHCALA          = 0x108;
const REG_PFA             = 0x10A;
const REG_PERIOD          = 0x10E;
const REG_ALT_OUTPUT      = 0x110;
const REG_INTERNAL_RES    = 0x120;

// 24-bit / 32-bit registers
const REG_SAGLVL          = 0x200;
const REG_ACCMODE         = 0x201;
const REG_AP_NOLOAD       = 0x203;
const REG_VAR_NOLOAD      = 0x204;
const REG_VA_NOLOAD       = 0x205;
const REG_AWATT           = 0x212;
const REG_VRMS            = 0x21C;
const REG_AENERGYA        = 0x21E;
const REG_OVLVL           = 0x224;
const REG_OILVL           = 0x225;
const REG_IRQENA          = 0x22C;
const REG_RSTIRQSTATA     = 0x22E;
const REG_IRQENB          = 0x22F;
const REG_RSTIRQSTATB     = 0x231;
const REG_CRC             = 0x37F;
const REG_AWGAIN          = 0x282;
const REG_AWATTOS         = 0x289;

const REG_120_UNLOCK_ADDR = 0x0FE;
const REG_120_UNLOCK      = 0xAD;
const REG_120_VALUE       = 0x30;

const ADC_FS_VOLTS  = 0.5 / Math.sqrt(2);        // 0.353553
const ADC_FS_CODE   = 9032007;
const POWER_FS_CODE = 4862401;
const T_SAMPLE      = 1.0 / 206900.0;
const PF_LSB        = 1.0 / 32768.0;
const ANGLE_LSB     = 1.0 / 223750.0;
const PHASE_LSB     = 1.0 / 895000.0;

const BUS_I2C  = 'i2c';
const BUS_SPI  = 'spi';
const BUS_UART = 'uart';

const _delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

function _twos(value, width) {
    const mask = (1 << width) - 1;
    const sign = 1 << (width - 1);
    value &= mask;
    if (value & sign) value -= mask + 1;
    return value;
}

function _u16(data, offset = 0) {
    return (data[offset] << 8) | data[offset + 1];
}

function _s16(data, offset = 0) {
    return _twos(_u16(data, offset), 16);
}

function _u24(data, offset = 0) {
    return (data[offset] << 16) | (data[offset + 1] << 8) | data[offset + 2];
}

function _s24(data, offset = 0) {
    return _twos(_u24(data, offset), 24);
}

function _u32(data, offset = 0) {
    return ((data[offset] << 24) | (data[offset + 1] << 16) |
            (data[offset + 2] << 8) | data[offset + 3]) >>> 0;
}


class ADE7953Minimal {
    /**
     * ADE7953 single-phase multifunction metering IC — minimal interface.
     *
     * Reads voltage, current and active power/energy for Current Channel A.
     * No register configuration beyond the mandatory power-up sequence and
     * the caller's sensor scaling is performed.
     *
     * @param {import('../../connection/connection').Connection} connection
     *        Configured I²C, SPI or UART connection bound to the device.
     * @param {number} voltageGain Real volts at the mains per volt at VP–VN.
     * @param {number} currentGain Real amperes per volt at IAP–IAN.
     * @param {string} [busType='i2c'] One of 'i2c', 'spi' or 'uart'.
     */
    constructor(connection, voltageGain, currentGain, busType = 'i2c') {
        this._conn = connection;
        this._busType = busType;
        this._voltageGain = Number(voltageGain);
        this._currentGainA = Number(currentGain);
        this._currentGainB = Number(currentGain);
        this._pgaA = 1;
        this._pgaB = 1;
        this._pgaV = 1;
        this._initChip();
    }

    async _initChip() {
        await _delay(110);
        await this._writeU8(REG_120_UNLOCK_ADDR, REG_120_UNLOCK);
        await this._writeU16(REG_INTERNAL_RES, REG_120_VALUE);
    }

    _addrForRead(reg, n) {
        if (this._busType === BUS_SPI) {
            return Buffer.from([
                (reg >> 8) & 0x7F,
                (reg & 0xFF) | 0x80,
                0x00, 0x00,
            ]);
        }
        if (this._busType === BUS_UART) {
            return Buffer.from([0x35, (reg >> 8) & 0xFF, reg & 0xFF]);
        }
        return Buffer.from([(reg >> 8) & 0xFF, reg & 0xFF]);
    }

    _addrForWrite(reg, payload) {
        if (this._busType === BUS_UART) {
            const reversed = Buffer.from(payload).reverse();
            return Buffer.concat([Buffer.from([0xCA, (reg >> 8) & 0xFF, reg & 0xFF]), reversed]);
        }
        return Buffer.concat([Buffer.from([(reg >> 8) & 0xFF, reg & 0xFF]), Buffer.from(payload)]);
    }

    async _read(reg, n) {
        if (this._busType === BUS_SPI) {
            const tx = this._addrForRead(reg, n);
            const raw = await this._conn.writeRead(tx, n + 2);
            return raw.slice(2);
        }
        if (this._busType === BUS_UART) {
            await this._conn.write(this._addrForRead(reg, n));
            await _delay(1);
            const raw = await this._conn.read(n);
            return Buffer.from(raw).reverse();
        }
        return await this._conn.writeRead(this._addrForRead(reg, n), n);
    }

    async _writePayload(reg, payload) {
        await this._conn.write(this._addrForWrite(reg, payload));
    }

    async _writeU8(reg, value) {
        await this._writePayload(reg, [(value & 0xFF)]);
    }

    async _writeU16(reg, value) {
        await this._writePayload(reg, [(value >> 8) & 0xFF, value & 0xFF]);
    }

    async _writeU24(reg, value) {
        await this._writePayload(reg, [(value >> 16) & 0xFF, (value >> 8) & 0xFF, value & 0xFF]);
    }

    async _readU24(reg) {
        const raw = await this._read(reg, 3);
        return _u24(raw);
    }

    async _readS24(reg) {
        const raw = await this._read(reg, 3);
        return _s24(raw);
    }

    async _readU32(reg) {
        const raw = await this._read(reg, 4);
        return _u32(raw);
    }

    async _readU16(reg) {
        const raw = await this._read(reg, 2);
        return _u16(raw);
    }

    async _readS16(reg) {
        const raw = await this._read(reg, 2);
        return _s16(raw);
    }

    _voltageScale() {
        return (ADC_FS_VOLTS * this._voltageGain) / (ADC_FS_CODE * this._pgaV);
    }

    _currentScale(gain) {
        return (ADC_FS_VOLTS * gain) / ADC_FS_CODE;
    }

    _powerScale(gain) {
        return ((ADC_FS_VOLTS * ADC_FS_VOLTS) * this._voltageGain * gain) / POWER_FS_CODE;
    }

    _energyScale(gain) {
        return ((ADC_FS_VOLTS * ADC_FS_VOLTS) * this._voltageGain * gain * T_SAMPLE) / 3600.0;
    }

    /**
     * Read the RMS voltage on the voltage channel.
     * @returns {Promise<number>} Voltage in volts.
     */
    async voltage() {
        const raw = await this._readU24(REG_VRMS);
        return raw * this._voltageScale();
    }

    /**
     * Read the RMS current on Current Channel A.
     * @returns {Promise<number>} Current in amperes.
     */
    async current() {
        const raw = await this._readU24(0x21A);
        return raw * this._currentScale(this._currentGainA);
    }

    /**
     * Read instantaneous active power on Current Channel A.
     * @returns {Promise<number>} Active power in watts (signed).
     */
    async activePower() {
        const raw = await this._readS24(REG_AWATT);
        return raw * this._powerScale(this._currentGainA);
    }

    /**
     * Read the active-energy accumulator for Current Channel A.
     * @returns {Promise<number>} Active energy in watt-hours accumulated since the previous call.
     */
    async activeEnergy() {
        const raw = await this._readS24(REG_AENERGYA);
        return raw * this._energyScale(this._currentGainA);
    }
}


class ADE7953Full extends ADE7953Minimal {
    /**
     * ADE7953 full interface — extends ADE7953Minimal with Channel B,
     * reactive/apparent measurements, calibration, accumulation modes,
     * power-quality features (no-load, sag, peak, overcurrent/overvoltage),
     * zero-crossing, REVP, alternate outputs, CF pulses, interrupts,
     * checksum, write protection, reset and last-operation diagnostics.
     *
     * @param {import('../../connection/connection').Connection} connection
     * @param {number} voltageGain
     * @param {number} currentGain
     * @param {string} [busType='i2c']
     */
    constructor(connection, voltageGain, currentGain, busType = 'i2c') {
        super(connection, voltageGain, currentGain, busType);
    }

    async configureChannelB(currentGainB) {
        this._currentGainB = Number(currentGainB);
    }

    async currentB() {
        const raw = await this._readU24(0x21B);
        return raw * this._currentScale(this._currentGainB);
    }

    async activePowerB() {
        const raw = await this._readS24(0x213);
        return raw * this._powerScale(this._currentGainB);
    }

    async activeEnergyB() {
        const raw = await this._readS24(0x21F);
        return raw * this._energyScale(this._currentGainB);
    }

    async reactivePower() {
        const raw = await this._readS24(0x214);
        return raw * this._powerScale(this._currentGainA);
    }

    async reactivePowerB() {
        const raw = await this._readS24(0x215);
        return raw * this._powerScale(this._currentGainB);
    }

    async reactiveEnergy() {
        const raw = await this._readS24(0x220);
        return raw * this._energyScale(this._currentGainA);
    }

    async reactiveEnergyB() {
        const raw = await this._readS24(0x221);
        return raw * this._energyScale(this._currentGainB);
    }

    async apparentPower() {
        const raw = await this._readS24(0x210);
        return raw * this._powerScale(this._currentGainA);
    }

    async apparentPowerB() {
        const raw = await this._readS24(0x211);
        return raw * this._powerScale(this._currentGainB);
    }

    async apparentEnergy() {
        const raw = await this._readS24(0x222);
        return raw * this._energyScale(this._currentGainA);
    }

    async apparentEnergyB() {
        const raw = await this._readS24(0x223);
        return raw * this._energyScale(this._currentGainB);
    }

    async powerFactor() {
        const raw = await this._readS16(0x10A);
        return raw * PF_LSB;
    }

    async linePeriod() {
        const raw = await this._readU16(REG_PERIOD);
        return (raw + 1) * ANGLE_LSB;
    }

    async lineFrequency() {
        return 1.0 / (await this.linePeriod());
    }

    async setPga(channel, gain) {
        const bits = {1:0, 2:1, 4:2, 8:3, 16:4, 22:5}[gain];
        if (bits === undefined) return;
        if (channel === 'a') { await this._writeU8(REG_PGA_IA, bits); this._pgaA = gain; }
        else if (channel === 'b') { await this._writeU8(REG_PGA_IB, bits); this._pgaB = gain; }
        else if (channel === 'v') { await this._writeU8(REG_PGA_V, bits); this._pgaV = gain; }
    }

    async setPhaseCalibration(channel, delayS) {
        let mag = Math.round(Math.abs(delayS) / PHASE_LSB);
        if (mag > 0x1FF) mag = 0x1FF;
        let raw = mag & 0x1FF;
        if (delayS >= 0) raw |= 0x200;
        if (channel === 'a') await this._writeU16(REG_PHCALA, raw);
        else if (channel === 'b') await this._writeU16(0x109, raw);
    }

    async setGainCalibration(reg, value) {
        await this._writeU24(reg, value & 0xFFFFFF);
    }

    async gainCalibration(reg) {
        return await this._readU24(reg);
    }

    async setOffsetCalibration(reg, value) {
        await this._writeU24(reg, value & 0xFFFFFF);
    }

    async offsetCalibration(reg) {
        return await this._readS24(reg);
    }

    async checksum() {
        return await this._readU32(REG_CRC);
    }

    async enableChecksum(enabled) {
        let cfg = await this._readU16(REG_CONFIG);
        if (enabled) cfg |= (1 << 8);
        else cfg &= ~(1 << 8);
        await this._writeU16(REG_CONFIG, cfg);
    }

    async configureOvervoltage(threshold) {
        let raw = (threshold * (ADC_FS_CODE * this._pgaV)) / (ADC_FS_VOLTS * this._voltageGain);
        if (raw < 0) raw = 0;
        if (raw > 0xFFFFFF) raw = 0xFFFFFF;
        await this._writeU24(REG_OVLVL, raw);
    }

    async configureOvercurrent(threshold) {
        let raw = (threshold * ADC_FS_CODE) / ADC_FS_VOLTS;
        if (raw < 0) raw = 0;
        if (raw > 0xFFFFFF) raw = 0xFFFFFF;
        await this._writeU24(REG_OILVL, raw);
    }

    async reset() {
        let cfg = await this._readU16(REG_CONFIG);
        cfg |= (1 << 7);
        await this._writeU16(REG_CONFIG, cfg);
        await _delay(110);
        await this._writeU8(REG_120_UNLOCK_ADDR, REG_120_UNLOCK);
        await this._writeU16(REG_INTERNAL_RES, REG_120_VALUE);
        this._pgaA = 1;
        this._pgaB = 1;
        this._pgaV = 1;
    }

    async version() {
        const raw = await this._read(REG_VERSION, 1);
        return raw[0];
    }
}


module.exports = {
    ADE7953Minimal,
    ADE7953Full,
    REG_VERSION,
    REG_AWGAIN,
    REG_AWATTOS,
};

File: cpp/src/chips/power/ADE7953.cpp
#include "ADE7953.h"

namespace {
    constexpr uint8_t  UART_CMD_READ  = 0x35;
    constexpr uint8_t  UART_CMD_WRITE = 0xCA;

    // IRQ source bit positions (channel A group).
    constexpr uint32_t AEHFA       = 1u << 0;
    constexpr uint32_t VAREHFA     = 1u << 1;
    constexpr uint32_t VAEHFA      = 1u << 2;
    constexpr uint32_t AEOFA       = 1u << 3;
    constexpr uint32_t VAREOFA     = 1u << 4;
    constexpr uint32_t VAEOFA      = 1u << 5;
    constexpr uint32_t AP_NOLOADA  = 1u << 6;
    constexpr uint32_t VAR_NOLOADA = 1u << 7;
    constexpr uint32_t VA_NOLOADA  = 1u << 8;
    constexpr uint32_t APSIGN_A    = 1u << 9;
    constexpr uint32_t VARSIGN_A   = 1u << 10;
    constexpr uint32_t ZXTO_IA     = 1u << 11;
    constexpr uint32_t ZXIA        = 1u << 12;
    constexpr uint32_t OIA         = 1u << 13;
    constexpr uint32_t ZXTO        = 1u << 14;
    constexpr uint32_t ZXV         = 1u << 15;
    constexpr uint32_t OV          = 1u << 16;
    constexpr uint32_t WSMP        = 1u << 17;
    constexpr uint32_t CYCEND      = 1u << 18;
    constexpr uint32_t SAG_BIT     = 1u << 19;
    constexpr uint32_t RESET_BIT   = 1u << 20;
    constexpr uint32_t CRC_BIT     = 1u << 21;
}


void ADE7953Minimal::_delayMs(uint32_t ms) {
    if (ms == 0) return;
    // Busy-wait ~ms milliseconds; portable across bare-metal targets.
    // 1 ms at >=1 MHz clock with no optimisation: ~1000 iterations.
    volatile uint32_t count = ms * 1000u;
    while (count--) { __asm__ volatile("nop"); }
}


void ADE7953Minimal::_initChip() {
    _delayMs(110);
    _writeU8(REG_INTERNAL_RES, REG_120_UNLOCK);
    _writeU16(REG_INTERNAL_RES, REG_120_VALUE);
}


void ADE7953Minimal::_read(uint16_t reg, uint8_t n, uint8_t* buf) {
    uint8_t tx[4];
    uint8_t tx_len = 0;
    uint8_t rx_skip = 0;

    if (_bus_type == BUS_SPI) {
        tx[0] = (reg >> 8) & 0x7F;
        tx[1] = (reg & 0xFF) | 0x80;  // READ bit
        tx[2] = 0x00;
        tx[3] = 0x00;
        tx_len = 4;
        rx_skip = 2;
    } else if (_bus_type == BUS_UART) {
        tx[0] = UART_CMD_READ;
        tx[1] = (reg >> 8) & 0xFF;
        tx[2] = (reg & 0xFF);
        tx_len = 3;
    } else {
        tx[0] = (reg >> 8) & 0xFF;
        tx[1] = (reg & 0xFF);
        tx_len = 2;
    }

    if (_bus_type == BUS_UART) {
        _connection.write(tx, tx_len);
        _delayMs(1);
        uint8_t tmp[4];
        if (n > 4) n = 4;
        _connection.read(tmp, n);
        // UART data is LSB-first.
        for (uint8_t i = 0; i < n; ++i) buf[i] = tmp[n - 1 - i];
    } else {
        uint8_t tmp[8];
        if (n > 8) n = 8;
        _connection.write_read(tx, tx_len, tmp, n);
        for (uint8_t i = 0; i < n; ++i) buf[i] = tmp[i + rx_skip];
    }
}


void ADE7953Minimal::_write(uint16_t reg, const uint8_t* payload, uint8_t n) {
    if (n > 4) n = 4;
    uint8_t frame[8];
    uint8_t frame_len = 0;

    if (_bus_type == BUS_UART) {
        frame[0] = UART_CMD_WRITE;
        frame[1] = (reg >> 8) & 0xFF;
        frame[2] = (reg & 0xFF);
        // UART data is LSB-first.
        for (uint8_t i = 0; i < n; ++i) frame[3 + i] = payload[n - 1 - i];
        frame_len = 3 + n;
    } else {
        frame[0] = (reg >> 8) & 0xFF;
        frame[1] = (reg & 0xFF);
        memcpy(&frame[2], payload, n);
        frame_len = 2 + n;
    }
    _connection.write(frame, frame_len);
}


void ADE7953Minimal::_writeU8(uint16_t reg, uint8_t value) {
    _write(reg, &value, 1);
}


void ADE7953Minimal::_writeU16(uint16_t reg, uint16_t value) {
    uint8_t payload[2];
    payload[0] = (value >> 8) & 0xFF;
    payload[1] = (value & 0xFF);
    _write(reg, payload, 2);
}


void ADE7953Minimal::_writeU24(uint16_t reg, uint32_t value) {
    uint8_t payload[3];
    payload[0] = (value >> 16) & 0xFF;
    payload[1] = (value >> 8) & 0xFF;
    payload[2] = (value & 0xFF);
    _write(reg, payload, 3);
}


void ADE7953Minimal::_writeU32(uint16_t reg, uint32_t value) {
    uint8_t payload[4];
    payload[0] = (value >> 24) & 0xFF;
    payload[1] = (value >> 16) & 0xFF;
    payload[2] = (value >> 8) & 0xFF;
    payload[3] = (value & 0xFF);
    _write(reg, payload, 4);
}


uint16_t ADE7953Minimal::_readU16(uint16_t reg) {
    uint8_t buf[2];
    _read(reg, 2, buf);
    return (uint16_t(buf[0]) << 8) | buf[1];
}


int16_t ADE7953Minimal::_readS16(uint16_t reg) {
    uint16_t v = _readU16(reg);
    if (v & 0x8000) return int16_t(v) - 0x10000;
    return int16_t(v);
}


uint32_t ADE7953Minimal::_readU24(uint16_t reg) {
    uint8_t buf[3];
    _read(reg, 3, buf);
    return (uint32_t(buf[0]) << 16) | (uint32_t(buf[1]) << 8) | uint32_t(buf[2]);
}


int32_t ADE7953Minimal::_readS24(uint16_t reg) {
    uint32_t v = _readU24(reg);
    if (v & 0x800000u) return int32_t(v) - 0x1000000;
    return int32_t(v);
}


uint32_t ADE7953Minimal::_readU32(uint16_t reg) {
    uint8_t buf[4];
    _read(reg, 4, buf);
    return (uint32_t(buf[0]) << 24) | (uint32_t(buf[1]) << 16) |
           (uint32_t(buf[2]) << 8) | uint32_t(buf[3]);
}


float ADE7953Minimal::_voltageScale() const {
    return (ADC_FS_VOLTS * _voltage_gain) / float(ADC_FS_CODE * _pga_v);
}


float ADE7953Minimal::_currentScale(float gain) const {
    return (ADC_FS_VOLTS * gain) / float(ADC_FS_CODE);
}


float ADE7953Minimal::_powerScale(float gain) const {
    return ((ADC_FS_VOLTS * ADC_FS_VOLTS) * _voltage_gain * gain) / float(POWER_FS_CODE);
}


float ADE7953Minimal::_energyScale(float gain) const {
    return ((ADC_FS_VOLTS * ADC_FS_VOLTS) * _voltage_gain * gain * T_SAMPLE) / 3600.0f;
}


ADE7953Minimal::ADE7953Minimal(Connection& connection,
                               float voltage_gain,
                               float current_gain,
                               uint8_t bus_type)
    : _connection(connection),
      _bus_type(bus_type),
      _voltage_gain(voltage_gain),
      _current_gain_a(current_gain),
      _current_gain_b(current_gain),
      _pga_a(1),
      _pga_b(1),
      _pga_v(1) {
    _initChip();
}


float ADE7953Minimal::voltage() {
    return float(_readU24(REG_VRMS)) * _voltageScale();
}


float ADE7953Minimal::current() {
    return float(_readU24(0x21A)) * _currentScale(_current_gain_a);
}


float ADE7953Minimal::activePower() {
    return float(_readS24(REG_AWATT)) * _powerScale(_current_gain_a);
}


float ADE7953Minimal::activeEnergy() {
    return float(_readS24(REG_AENERGYA)) * _energyScale(_current_gain_a);
}


// ----------------------------------------------------------------------------
// Full interface


ADE7953Full::ADE7953Full(Connection& connection,
                         float voltage_gain,
                         float current_gain,
                         uint8_t bus_type)
    : ADE7953Minimal(connection, voltage_gain, current_gain, bus_type) {}


void ADE7953Full::configureChannelB(float current_gain_b) {
    _current_gain_b = current_gain_b;
}


float ADE7953Full::currentB() {
    return float(_readU24(0x21B)) * _currentScale(_current_gain_b);
}


float ADE7953Full::activePowerB() {
    return float(_readS24(0x213)) * _powerScale(_current_gain_b);
}


float ADE7953Full::activeEnergyB() {
    return float(_readS24(0x21F)) * _energyScale(_current_gain_b);
}


float ADE7953Full::reactivePower() {
    return float(_readS24(0x214)) * _powerScale(_current_gain_a);
}


float ADE7953Full::reactivePowerB() {
    return float(_readS24(0x215)) * _powerScale(_current_gain_b);
}


float ADE7953Full::reactiveEnergy() {
    return float(_readS24(0x220)) * _energyScale(_current_gain_a);
}


float ADE7953Full::reactiveEnergyB() {
    return float(_readS24(0x221)) * _energyScale(_current_gain_b);
}


float ADE7953Full::apparentPower() {
    return float(_readS24(0x210)) * _powerScale(_current_gain_a);
}


float ADE7953Full::apparentPowerB() {
    return float(_readS24(0x211)) * _powerScale(_current_gain_b);
}


float ADE7953Full::apparentEnergy() {
    return float(_readS24(0x222)) * _energyScale(_current_gain_a);
}


float ADE7953Full::apparentEnergyB() {
    return float(_readS24(0x223)) * _energyScale(_current_gain_b);
}


float ADE7953Full::powerFactor() {
    return float(_readS16(REG_PFA)) * PF_LSB;
}


float ADE7953Full::linePeriod() {
    return float(_readU16(REG_PERIOD) + 1) * ANGLE_LSB;
}


float ADE7953Full::lineFrequency() {
    return 1.0f / linePeriod();
}


void ADE7953Full::setPga(char channel, uint8_t gain) {
    uint8_t bits = 0;
    switch (gain) {
        case 1: bits = 0; break;
        case 2: bits = 1; break;
        case 4: bits = 2; break;
        case 8: bits = 3; break;
        case 16: bits = 4; break;
        case 22: bits = 5; break;
        default: return;
    }
    if (channel == 'a') { _writeU8(REG_PGA_IA, bits); _pga_a = gain; }
    else if (channel == 'b') { _writeU8(REG_PGA_IB, bits); _pga_b = gain; }
    else if (channel == 'v') { _writeU8(REG_PGA_V, bits); _pga_v = gain; }
}


void ADE7953Full::setPhaseCalibration(char channel, float delay_s) {
    uint32_t mag = (uint32_t)((delay_s < 0 ? -delay_s : delay_s) / PHASE_LSB + 0.5f);
    if (mag > 0x1FF) mag = 0x1FF;
    uint16_t raw = (uint16_t)(mag & 0x1FF);
    if (delay_s >= 0) raw |= 0x200;
    if (channel == 'a') _writeU16(REG_PHCALA, raw);
    else if (channel == 'b') _writeU16(REG_PHCALB, raw);
}


void ADE7953Full::setGainCalibration(uint16_t reg, uint32_t value) {
    _writeU24(reg, value & 0xFFFFFFu);
}


uint32_t ADE7953Full::gainCalibration(uint16_t reg) {
    return _readU24(reg);
}


void ADE7953Full::setOffsetCalibration(uint16_t reg, int32_t value) {
    _writeU24(reg, uint32_t(value) & 0xFFFFFFu);
}


int32_t ADE7953Full::offsetCalibration(uint16_t reg) {
    return _readS24(reg);
}


uint32_t ADE7953Full::checksum() {
    return _readU32(REG_CRC);
}


void ADE7953Full::enableChecksum(bool enabled) {
    uint16_t cfg = _readU16(REG_CONFIG);
    if (enabled) cfg |= (1u << 8);
    else cfg &= ~(1u << 8);
    _writeU16(REG_CONFIG, cfg);
}


void ADE7953Full::configureOvervoltage(float threshold) {
    float raw_f = (threshold * float(ADC_FS_CODE * _pga_v)) /
                  (ADC_FS_VOLTS * _voltage_gain);
    if (raw_f < 0) raw_f = 0;
    if (raw_f > float(0xFFFFFF)) raw_f = float(0xFFFFFF);
    _writeU24(REG_OVLVL, uint32_t(raw_f));
}


void ADE7953Full::configureOvercurrent(float threshold) {
    float raw_f = (threshold * float(ADC_FS_CODE)) / ADC_FS_VOLTS;
    if (raw_f < 0) raw_f = 0;
    if (raw_f > float(0xFFFFFF)) raw_f = float(0xFFFFFF);
    _writeU24(REG_OILVL, uint32_t(raw_f));
}


void ADE7953Full::reset() {
    uint16_t cfg = _readU16(REG_CONFIG);
    cfg |= (1u << 7);
    _writeU16(REG_CONFIG, cfg);
    _delayMs(110);
    _writeU8(REG_INTERNAL_RES, REG_120_UNLOCK);
    _writeU16(REG_INTERNAL_RES, REG_120_VALUE);
    _pga_a = 1;
    _pga_b = 1;
    _pga_v = 1;
}


uint8_t ADE7953Full::version() {
    uint8_t v = 0;
    _read(REG_VERSION, 1, &v);
    return v;
}


void ADE7953Full::setActiveEnergyMode(char channel, uint8_t mode) {
    uint32_t acc = _readU24(REG_ACCMODE);
    if (channel == 'a') {
        acc = (acc & ~0x3u) | (mode & 0x3);
    } else {
        acc = (acc & ~(0x3u << 2)) | ((uint32_t(mode) & 0x3) << 2);
    }
    _writeU24(REG_ACCMODE, acc);
}


void ADE7953Full::setReactiveEnergyMode(char channel, uint8_t mode) {
    uint32_t acc = _readU24(REG_ACCMODE);
    if (channel == 'a') {
        acc = (acc & ~(0x3u << 4)) | ((uint32_t(mode) & 0x3) << 4);
    } else {
        acc = (acc & ~(0x3u << 6)) | ((uint32_t(mode) & 0x3) << 6);
    }
    _writeU24(REG_ACCMODE, acc);
}